stdlib/json.gr

```diff
@@ -0,0 +1,2103 @@
+ * }
+ * ```
+ *
+ * ```
+ * {↵
+ * ··"a":·1,↵
+ * ··"b":·2,↵
+ * ··"c":·3↵
+ * }
+ * ```
+ */
+provide enum ObjectFormat {
+  CompactObjectEntries,
+  SpacedObjectEntries,
+  OneObjectEntryPerLine,
+}
+
+/**
+ * Controls line ending type in custom formatting.
+ */
+provide enum LineEnding {
+  NoLineEnding,
+  LineFeed,
+  CarriageReturnLineFeed,
+  CarriageReturn,
+}
+
+/*
+ * Allows fine-grained control of formatting in JSON printing.
+ */
+record FormattingSettings {
+  indentation: IndentationFormat,
+  arrayFormat: ArrayFormat,
+  objectFormat: ObjectFormat,
+  lineEnding: LineEnding,
+  finishWithNewLine: Bool,
+  escapeAllControlPoints: Bool,
+  escapeHTMLUnsafeSequences: Bool,
+  escapeNonASCII: Bool,
+}
+
+/**
+ * Allows control of formatting in JSON printing.
+ */
+provide enum FormattingChoices {
+  /**
+   * Recommended human readable formatting.
+   *
+   * Escapes all control points for the sake of clarity, but prints unicode
+   * codepoints directly so the result needs to be treated as proper unicode and
+   * is not safe to be transported in ASCII encoding.
+   *
+   * Roughly Equivalent to:
+   * ```
+   * Custom{
+   *  indentation: IndentWithSpaces(2),
+   *  arrayFormat: OneArrayEntryPerLine,
+   *  objectFormat: OneObjectEntryPerLine,
+   *  lineEnding: LineFeed,
+   *  finishWithNewLine: true,
+   *  escapeAllControlPoints: true,
+   *  escapeHTMLUnsafeSequences: false,
+   *  escapeNonASCII: false,
+   * }
+   * ```
+   *
+   * The following example have whitespaces, line breaks and control points
+   * replaced with visible characters.
+   * ```
+   * {↵
+   * ··"currency":·"€",↵
+   * ··"price":·99.9,↵
+   * ··"currencyDescription":·"EURO\u007f",↵
+   * ··"script·unembeddable":·"You·cannot·paste·a·JSON·object·directly·into·a·<script>·tag·in·HTML·if·it·contains·</·with·unescaped·forward·slash"↵
+   * }
+   * ```
+   *
+   * @since v0.6.0
+   */
+  Pretty,
+  /**
+   * Compact formatting that minimizes the size of resulting JSON at cost of not
+   * being easily human readable.
+   *
+   * Only performs minimal string escaping as required by the ECMA-404 standard,
+   * so the result needs to be treated as proper unicode and is not safe to be
+   * transported in ASCII encoding.
+   *
+   * Roughly Equivalent to:
+   * ```
+   * Custom{
+   *  indentation: NoIndentation,
+   *  arrayFormat: CompactArrayEntries,
+   *  objectFormat: CompactObjectEntries,
+   *  lineEnding: NoLineEnding,
+   *  finishWithNewLine: false,
+   *  escapeAllControlPoints: false,
+   *  escapeHTMLUnsafeSequences: false,
+   *  escapeNonASCII: false,
+   * }
+   * ```
+   *
+   * The following example have whitespaces, line breaks and control points
+   * replaced with visible characters.
+   * ```
+   * {"currency":"€","price":99.9,"currencyDescription":"EURO␡","script·unembeddable":"You·cannot·paste·a·JSON·object·directly·into·a·<script>·tag·in·HTML·if·it·contains·</·with·unescaped·forward·slash"}
+   * ```
+   *
+   * @since v0.6.0
```

Comment:
Does this attribute work on variants?

Reply:
https://github.com/grain-lang/grain/pull/1852/files#r1468986721 looks like it.